src/ImageSharp/Formats/Jpeg/Components/Decoder/HuffmanScanDecoder.cs

```diff
@@ -704,355 +568,302 @@ private unsafe void ParseProgressiveDataNonInterleaved()
+            /// <summary>
+            /// Gets or sets the current, if any, marker in the input stream.
+            /// </summary>
+            public byte Marker { get; set; }
 
-        [MethodImpl(InliningOptions.ShortMethod)]
-        private bool ContinueOnMcuComplete()
-        {
-            if (--this.todo > 0)
+            /// <summary>
+            /// Gets or sets the opening position of an identified marker.
+            /// </summary>
+            public long MarkerPosition { get; set; }
+
+            /// <summary>
+            /// Gets or sets a value indicating whether we have a bad marker, I.E. One that is not between RST0 and RST7
+            /// </summary>
+            public bool BadMarker { get; set; }
+
+            /// <summary>
+            /// Gets or sets the entropy encoded code buffer.
+            /// </summary>
+            public ulong Data { get; set; }
+
+            /// <summary>
+            /// Gets or sets the number of valid bits left to read in the buffer.
+            /// </summary>
+            public int Remain { get; set; }
+
+            /// <summary>
+            /// Gets or sets a value indicating whether there is more data to pull from the stream for the current mcu.
+            /// </summary>
+            public bool NoMore { get; set; }
+
+            /// <summary>
+            /// Gets or sets a value indicating whether we have prematurely reached the end of the file.
+            /// </summary>
+            public bool Eof { get; set; }
+
+            [MethodImpl(InliningOptions.ShortMethod)]
+            public void CheckBits()
             {
-                return true;
+                if (this.Remain < 16)
+                {
+                    this.FillBuffer();
+                }
             }
 
-            if (this.codeBits < 24)
+            [MethodImpl(InliningOptions.ShortMethod)]
+            public void Reset()
             {
-                this.FillBuffer();
+                this.Data = 0ul;
+                this.Remain = 0;
+                this.Marker = JpegConstants.Markers.XFF;
+                this.MarkerPosition = 0;
+                this.BadMarker = false;
+                this.NoMore = false;
+                this.Eof = false;
             }
 
-            // If it's NOT a restart, then just bail, so we get corrupt data rather than no data.
-            // Reset the stream to before any bad markers to ensure we can read successive segments.
-            if (this.badMarker)
+            [MethodImpl(InliningOptions.ShortMethod)]
+            public bool HasRestart()
             {
-                this.stream.Position = this.markerPosition;
+                byte m = this.Marker;
+                return m >= JpegConstants.Markers.RST0 && m <= JpegConstants.Markers.RST7;
             }
 
-            if (!this.HasRestart())
+            [MethodImpl(InliningOptions.ShortMethod)]
```

Comment:
Based on your previous discussion, my understanding is that this should be `ColdPath`, and `GetBytes` should be `ShortMethod`

Reply:
Gone for `ShortMethod` for now on both. Everything I've attempted to improve stream reading performance above and beyond the buffered stream results in worse performance.